<a href="https://colab.research.google.com/github/kleeresearch/TextAnalysis/blob/master/Course/Day3/11_vector_example.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### 벡터 생성하기

In [2]:
import numpy as np # numpy를 사용해서 벡터와 행렬 생성

In [3]:
a = np.array([1,2]) # array()를 이용하는 경우는 보통 리스트를 인자로 입력

In [4]:
a

array([1, 2])

In [5]:
a.shape # 아래 결과는 종벡터를 의미

(2,)

In [6]:
len(a)

2

### 거리 구하기

a가 원점으로부터의 길이가 유클리디안 거리가 됨.

a = (1,2)
O = (0,0)

a = root ((1-0)^2 + (2-0)^2 ) = root 5


In [7]:
# 벡터 a의 길이 구하기
np.linalg.norm(a)

2.23606797749979

#norm() 함수를 이용해서 두 벡터 사이의 유클리디안 거리 구하기

a + b = (a1+b1, a2+b2) = (1+2, 2+2) = (3,4)

벡터끼리의 합은 또 다른 벡터가 된다.

a - b = (a1-b1, a2-b2)

벡터끼리의 차는 또 다른 벡터가 된다.

norm(z) = 원점으로부터의 유클리디안 거리는 같은 위치의 벡터끼리 빼 다음에 제곱한다.

norm(z) = root[ (y1-x1-o)^2 + (y2-x2-o)^2 ]

x 와 y 의  차이로 표현되는 벡터z 의 길이는 x 와 y 사이로 표현되는 유클리디안 거리와 똑같다.

norm 펑션을 가지고 유클리디안 거리를 구하기 위해서는 norm(y-x)로 계산하면 된다.

norm(b-a) = ab = 1  두 요소 간의 유클리디안 거리와 같다.

norm(c-a) = root((-3-1)^2 +(-3-2)^2)) = root(41) = 6.xxxx



In [8]:
# 3 개 벡터 만들기
a = np.array([1,2])
b = np.array([2,2])
c = np.array([-3, -3])

In [31]:
print(np.linalg.norm(b-a))
print(np.linalg.norm(c-a))

1.0
6.4031242374328485


6.403은 root41 을 의미함.

### scipy 모듈에서 제공되는 거리 함수 사용하기

In [10]:
import scipy.spatial.distance as dst

시티블록 거리 (L1 norm) 계산하기

In [11]:
dst.cityblock(a,b)

1

In [12]:
dst.cityblock(a,c)

9

linalg 모듈의 norm() 함수를 이용한 L1-norm 거리 계산

In [13]:
np.linalg.norm(c-a, 1)

9.0

distance 모듈에서 제공되는 함수를 이용한 유클리디안 거리 계산

In [14]:
dst.euclidean(a,b)

1.0

In [15]:
dst.euclidean(a,c)

6.4031242374328485

norm() 함수의 두 번째 인자로 2를 입력하는 경우의 예

In [16]:
np.linalg.norm(c-a, 2)

6.4031242374328485

In [17]:
np.linalg.norm(c-a)

6.4031242374328485

Jaccard distance

In [18]:
d = np.array([1,0,0])
e = np.array([1,1,0])

In [19]:
dst.jaccard(d,e)

0.5

Hamming distance

In [20]:
dst.hamming(d,e)

0.3333333333333333

# Cosine similarity

In [21]:
Doc1 = np.array([2,1,0,0,0,1])
Doc2 = np.array([1,0,2,1,0,0])
Doc3 = np.array([0,1,0,0,1,2])

In [22]:
print(np.linalg.norm(Doc2-Doc1))
print(np.linalg.norm(Doc3-Doc1))

2.8284271247461903
2.449489742783178


In [23]:
print(np.dot(Doc1,Doc2)/(np.linalg.norm(Doc1)*np.linalg.norm(Doc2)))
print(np.dot(Doc1,Doc3)/(np.linalg.norm(Doc1)*np.linalg.norm(Doc3)))

0.33333333333333337
0.5000000000000001


#### 내적 (dot product)

In [24]:
a = np.array([1,2])
b = np.array([2,2])
np.dot(a,b)

6

In [25]:
np.dot(a,b)/(np.linalg.norm(a)*np.linalg.norm(b))

0.9486832980505138

#### cosine distance

In [26]:
dst.cosine(a,b)

0.05131670194948623

In [27]:
1-np.dot(a,b)/(np.linalg.norm(a)*np.linalg.norm(b))

0.05131670194948623

# 예제 데이터 사용해 보기

In [28]:
import pandas as pd
data = pd.read_csv('vector_example.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'vector_example.csv'

In [ ]:
data

In [29]:
data_np = data.values

NameError: name 'data' is not defined

In [30]:
data_np

NameError: name 'data_np' is not defined

In [ ]:
# 첫 번째 관측치에 대한 벡터
data_np[0]

array([30,  1], dtype=int64)

In [ ]:
# 첫 번째 벡터와 두 번째 벡터 간의 유클리디안 거리
np.linalg.norm(data_np[1]-data_np[0])

3.1622776601683795

In [ ]:
np.linalg.norm(data_np[2]-data_np[0])

34.655446902326915

In [ ]:
# 첫 번째 벡터와 두 번째 벡터 간의 코사인 세타
np.dot(data_np[0],data_np[1])/(np.linalg.norm(data_np[0])*np.linalg.norm(data_np[1]))

0.999629802932285

In [ ]:
# 첫 번째 벡터와 세 번째 벡터 간의 코사인 세타
np.dot(data_np[0],data_np[2])/(np.linalg.norm(data_np[0])*np.linalg.norm(data_np[2]))

0.9236469142237134